crates/wast/src/component/alias.rs

```diff
@@ -28,7 +28,7 @@ impl<'a, const CORE: bool> Parse<'a> for InlineExportAlias<'a, CORE> {
 }
 
 /// An alias to a component item.
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
```

Comment:
The use case is to improve testing for roundtrip when using `Orca`. Currently what I am doing is writing the output of the roundtrip to a new `.wat` file and trying to do a string comparison. This does not work in cases such as: 
```rust
(component
  (type (;0;) (func))
  (import "one" (func (;0;) (type 0)))
  (import "two" (value $v (;0;) string))
  (type (;1;)
    (instance
      (type (;0;)
        (instance
          (core type (;0;)
            (module
              (type (;0;) (func))
              (import "six" "a" (func (type 0)))
              (type (;1;) (func))
              (import "six" "b" (func (type 1)))
            )
          )
          (export (;0;) "five" (core module (type 0)))
        )
      )
      (export (;0;) "four" (instance (type 0)))
    )
  )
  (import "three" (instance (;0;) (type 1)))
  (export (;1;) "four" (value $v))
)
```
On doing the roundtrip, it becomes:
```rust
(component
  (type (;0;) (func))
  (type (;1;)
    (instance
      (type (;0;)
        (instance
          (core type (;0;)
            (module
              (type (;0;) (func))
              (import "six" "a" (func (type 0)))
              (type (;1;) (func))
              (import "six" "b" (func (type 1)))
            )
          )
          (export (;0;) "five" (core module (type 0)))
        )
      )
      (export (;0;) "four" (instance (type 0)))
    )
  )
  (import "one" (func (;0;) (type 0)))
  (import "two" (value $v (;0;) string))
  (import "three" (instance (;0;) (type 1)))
  (export (;1;) "four" (value $v))
)
```
All the statements belonging to a particular section are clubbed together, for example the `imports`.
A direct string comparison will not suffice. Hence, I was thinking I could load the output `.wat` and parse it and then directly compare my IR to the original `.wat` file.
Something like 
```rust
assert_eq!(original_component, round_trip_component);
```
Considering my IR just uses `wasmparser` structs and enums, I wanted to add `Eq` and `PartialEq` to enable this

Reply:
My guess is that even if you could compare the textual ASTs directly it still wouldn't work in this case. Clumping like-kinded-items together works for core wasm but doesn't work for components, and the two components you've given here are not semantically equal to each other in all cases. For example types can be defined after an import and then referred to in later imports, but they can't always be reordered.

Reply:
Ohh I see. Is there something you can do with the ordering of sections such that it can generate valid components?
Looking at the example you've mentioned, does encoding the types first and then the imports and exports work?

Reply:
It's more that components allow interleaving sections unlike core wasm. This interleaving can be required as sections can introduce items referred to in other sections meaning that there's a dependency order between items in different sections. That in turn means that grouping everything together in the same section is not always equivalent, which is where I'm pointing out that the two components you've outlined above both have different ASTs and aren't the same.

Reply:
I've updated my IR to add positional information to preserve the ordering of the sections. In light of this, does my previous idea make sense? I can implement a `Eq`, `PartialEq` trait that processes the sections in order of their appearance in `.wat` files but I would still like to do something like `ComponentAlias == ComponentAlias`.

Reply:
Personally I'm still a bit leery of `==` working on AST items because it feels like a bit of a footgun. If the goal is to equate two components would it be possible to encode both to binary, print both from binary, and equate the results of the printing? That's roughtly what the `roundtrip.rs` tests do in a few places and if things are different it tries to figure out the diff of where things went wrong.

Reply:
That's what I was initially doing but in some cases I get a slight formatting difference in the output or I get `global $hi` change to `global 0`.
There is also a case where `(global $g1 (import "foo" "bar") i32)` becomes `(import "foo" "bar" (global (;0;) i32))`.
In all these cases, it looks like the roundtrip is slightly changing the actual string/byte contents but it still generates valid wasm binaries, i.e., they pass `wasm-tools validate`.
I was looking at this as a sort of backup approach in case I cannot output the exact same binary after a roundtrip, and thought comparing the AST in this case would be something like semantically comparing the wasm binaries and not relying on the text output to be the exact same.

Reply:
Yeah I don't think you'd want to compare a hand-written string but what I think you can do is:

* Start with hand-written string S
* use `wast` to convert S to binary and use `wasmprinter` to convert that to text, creating S1
* Use your own tooling to round-trip S as desired creating S'
* Use `wast` to convert S' to binary and `wasmprinter` to convert to text, creating S2
* Comparing S1 and S2 should return the right result

That should handle differences suchas improt sugar and such. That wouldn't handle cases of `$hi` vs `0` as that's a property of your round-tripping tool perhaps losing name sections?

In any case I think that if the round-tripping is producing different results I believe the textual ASTs would also be different?

Reply:
I've fixed the name sections and what you suggested is similar to what I'm doing.
Steps of what I'm doing:
1. Read in `.wat` using `wat::parse_file` creating `S`.
2. Encode it in the roundtrip and print using `wasmprinter` getting `S1`.
3. Compare `S` and `S1`.

Currently I still see it doing changes such as:
```rust
(module
  (global $g1 (import "foo" "bar") i32)
  (global $g2 i32 (i32.const 5))
  (global $g3 (mut i64) (i64.const 100))
)
```
becomes
```rust
(module
  (import "foo" "bar" (global (;0;) i32))
  (global (;1;) i32 i32.const 5)
  (global (;2;) (mut i64) i64.const 100)
)
```
This is because `(global $g1 (import "foo" "bar") i32)` is in the `Import` Section and stored as an import. So when it encodes, it extracts it from my Imports the `.wat` representation slightly changes.
Another issue is:
```rust
(func $#func67<_ZN4core3ptr191drop_in_place$LT$std..sync..poison..PoisonError$LT$std..sync..mutex..MutexGuard$LT$alloc..vec..Vec$LT$$LP$spin_sdk..wit..wasi..io..poll..Pollable$C$core..task..wake..Waker$RP$$GT$$GT$$GT$$GT$17h4e207d90019cfe5eE> (@name "_ZN4core3ptr191drop_in_place$LT$std..sync..poison..PoisonError$LT$std..sync..mutex..MutexGuard$LT$alloc..vec..Vec$LT$$LP$spin_sdk..wit..wasi..io..poll..Pollable$C$core..task..wake..Waker$RP$$GT$$GT$$GT$$GT$17h4e207d90019cfe5eE") (;67;) (type 1) (param i32)
```
becomes
```rust
(func $"#func67 _ZN4core3ptr191drop_in_place$LT$std..sync..poison..PoisonError$LT$std..sync..mutex..MutexGuard$LT$alloc..vec..Vec$LT$$LP$spin_sdk..wit..wasi..io..poll..Pollable$C$core..task..wake..Waker$RP$$GT$$GT$$GT$$GT$17h4e207d90019cfe5eE" (@name "_ZN4core3ptr191drop_in_place$LT$std..sync..poison..PoisonError$LT$std..sync..mutex..MutexGuard$LT$alloc..vec..Vec$LT$$LP$spin_sdk..wit..wasi..io..poll..Pollable$C$core..task..wake..Waker$RP$$GT$$GT$$GT$$GT$17h4e207d90019cfe5eE") (;67;) (type 1) (param i32)
```
This is despite me adding all the custom name sections which fixed all the other name issues. The output is still a valid binary but it's not exactly the same.
In both these case, I believe the AST would be fine to compare with because it gets stored as another function or in the first case, goes into the Imports section.

I also see some formatting differences in modules like `unreachable)` becomes
```rust
unreachable
)
```
which might result in string comparisons failing.

I do see that there is [wat-desugar](https://webassembly.github.io/wabt/doc/wat-desugar.1.html) in the `wabt` toolkit but I believe this does not support the component model and I was not able to find something similar in `wasm-tools`.

Reply:
You mentioned that in step 1 you do "Read in .wat using wat::parse_file creating S.". To confirm that means S is a binary, right? Are you comparing binaries or strings? If strings are you comparing the original `*.wat` file itself? (if so you should test the original `*.wat` file round-tripped through `wast` and `wasmprinter` first)

> This is despite me adding all the custom name sections which fixed all the other name issues.

This looks like the issue from awhile ago with two `wasmprinter` versions. One used to do `$#funcNN<foo>` and the other does `$"#funcNN foo"`

> I do see that there is [wat-desugar](https://webassembly.github.io/wabt/doc/wat-desugar.1.html) in the wabt toolkit but I believe this does not support the component model and I was not able to find something similar in wasm-tools.

I don't have experience with `wat-desugar` itself but my guess is that it's roughly equivalent to `wasm-tools print foo.wat`. That will convert `foo.wat` to binary automatically and then print it, printing a "canonical" form of sorts after handling all the sugar during the text parsing.

> wasmprinter is now giving me the same output for both pre-roundtrip and roundtrip files

I wonder if this was perhaps the wasmprinter version difference?

Reply:
> if so you should test the original *.wat file round-tripped through wast and wasmprinter first

I think this was our issue. Thank you so much for your advice on this! We didn't know there was some sort of desugaring happening behind the scenes.

Reply:
Great! Does that mean that the `Eq` and `PartialEq` impls on the AST are no longer needed?

Reply:
Nope, I don't think so but can it be merged regardless? It might prove to be useful in the future

Reply:
I'd prefer to leave these out because they have a compile-time cost and often aren't the right tool for the job. If it's ok with you mind removing this to only have the impls on the wasmparser types?

Reply:
Do you mean removing the traits from all the `wast` files? If yes, that works! In that case, should I just manually implement `Eq` and `PartialEq` for those structs/enums?

Reply:
Oh I mean removing the changes to the `crates/wast` folder in this PR, so not adding `derive(PartialEq, Eq)` to the types.

Reply:
Mind posting a version of the PR with that problem? I was unaware of `wasmparser` types depending on `wast`, so what you're seeing might be a subtly different error?

Reply:
Apologies, looks like I was mistaken. I've removed them from `wast` types and pushed it here itself.